File: src/main/java/com/jcloisterzone/game/capability/WatchtowerCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.TileModifier;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.board.pointer.ScoreMeeplePositionsPointer;
import com.jcloisterzone.event.ExprItem;
import com.jcloisterzone.event.PointsExpression;
import com.jcloisterzone.event.ScoreEvent;
import com.jcloisterzone.event.ScoreEvent.ReceivedPoints;
import com.jcloisterzone.feature.*;
import com.jcloisterzone.figure.Follower;
import com.jcloisterzone.figure.Meeple;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.reducers.AddPoints;

import io.vavr.Predicates;
import io.vavr.Tuple2;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;

import java.util.ArrayList;

public class WatchtowerCapability  extends Capability<Void> {

    public static class WatchtowerModifier extends TileModifier {
        private int points;
        private String subject;

        public WatchtowerModifier(String type) {
            super("Watchtower:" + type);
            String[] tokens = type.split("/", 2);
            this.points = Integer.parseInt(tokens[0]);
            this.subject = tokens[1];
        }
    }

    @Override
    public GameState beforeCompletableScore(GameState state, java.util.Set<Completable> features) {
        java.util.Map<Position, WatchtowerModifier> watchtowers = new java.util.HashMap<>();
        java.util.Map<Position, java.util.List<Tuple2<Meeple, FeaturePointer>>> watchtowerMeeples = new java.util.HashMap<>();

        state.getPlacedTiles().forEach((pos, pt) -> {
            pt.getTile().getTileModifiers().filter(Predicates.instanceOf(WatchtowerModifier.class)).forEach(mod -> {
                watchtowers.put(pos, (WatchtowerModifier) mod);
            });
        });

        if (watchtowers.isEmpty()) {
            return state;
        }

        for (Tuple2<Meeple, FeaturePointer> t : state.getDeployedMeeples()) {
            FeaturePointer fp = t._2;
            if (!watchtowers.containsKey(fp.getPosition())) {
                continue;
            }

            if (features.stream().filter(f -> f.getPlaces().contains(fp)).findAny().isPresent()) {
            	java.util.List<Tuple2<Meeple, FeaturePointer>> meeples = watchtowerMeeples.get(fp.getPosition());
                if (meeples == null) {
                    meeples = new ArrayList<>();
                    watchtowerMeeples.put(fp.getPosition(), meeples);
                }
                meeples.add(t);
            }
        }

        for (java.util.Map.Entry<Position, java.util.List<Tuple2<Meeple, FeaturePointer>>> entry : watchtowerMeeples.entrySet()) {
            Position pos = entry.getKey();
            WatchtowerModifier watchtower = watchtowers.get(pos);
            int count = 0;
            String exprName = null;

            GameState _state = state;
        	Stream<PlacedTile> pts = _state
                	.getAdjacentAndDiagonalTiles(pos)
                	.append(_state.getPlacedTile(pos));
        	
        	Set<Position> positions = pts.map(t -> t.getPosition()).toSet();

        	switch (watchtower.subject) {
                case "coat-of-arms":
                	for(PlacedTile pt: pts) {
                	  for (Feature f: pt.getTile().getInitialFeatures().values()) {
                        if (f instanceof City) {
                           	FeaturePointer fp = ((City) f).getPlaces().get().setPosition(pt.getPosition()).rotateCW(pt.getRotation());
                            if (!((City) _state.getFeature(fp)).hasModifier(_state, City.ELIMINATED_PENNANTS)) {
                              count+=((City) f).getModifier(_state, City.PENNANTS, 0);
                            }
                        }
                      }
                	}
                    exprName = "pennants";
                    break;
                case "monastery":
                    count = getNeigbouringFeatures(state, pos).filter(Predicates.instanceOf(Monastery.class)).length();
                    exprName = "monasteries";
                    break;
                case "city":
                    count = getNeigbouring(state, pos).filter(pt -> _state.getFeatureMap().get(pt.getPosition()).get().keySet().find(fp -> fp.getFeature().equals(City.class)).isDefined()).length();
                    exprName = "cities";
                    break;
                case "road":
                    count = getNeigbouring(state, pos).filter(pt -> _state.getFeatureMap().get(pt.getPosition()).get().keySet().find(fp -> fp.getFeature().equals(Road.class)).isDefined()).length();
                    exprName = "roads";
                    break;
                case "meeple":
                    count = state.getDeployedMeeples().filter((m, fp) -> {
                        if (!(m instanceof Follower)) return false;
                        Position mpos = fp.getPosition();
                        return Math.abs(pos.x - mpos.x) <= 1 && Math.abs(pos.y - mpos.y) <= 1;
                    }).length();
                    exprName = "meeples";
            }

            if (count > 0) {
                PointsExpression expr = new PointsExpression("watchtower", new ExprItem(count, exprName, count * watchtower.points));
                List<ReceivedPoints> receivedPoints = List.ofAll(entry.getValue()).map(t -> new ScoreEvent.ReceivedPoints(expr, t._1.getPlayer(), new ScoreMeeplePositionsPointer(t._1.getDeployment(_state), t._1.getId(), positions)));
                state = (new AddPoints(receivedPoints, true)).apply(state);
            }
        }

        return state;
    }

    private Stream<PlacedTile> getNeigbouring(GameState state, Position pos) {
        return state
                .getAdjacentAndDiagonalTiles(pos)
                .append(state.getPlacedTile(pos));
    }

    private Stream<Feature> getNeigbouringFeatures(GameState state, Position pos) {
        return getNeigbouring(state, pos).flatMap(pt -> pt.getTile().getInitialFeatures().values());
    }
}

File: src/main/ts/com/jcloisterzone/game/capability/WatchtowerCapability.ts
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import { List } from "../../../../io/vavr/SeqTypes.js";
import type { Tuple2 } from "../../../../io/vavr/Tuple.js";
import { Position } from "../../board/Position.js";
import { TileModifier } from "../../board/TileModifier.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import { ScoreMeeplePositionsPointer } from "../../board/pointer/ScoreMeeplePositionsPointer.js";
import { ExprItem } from "../../event/ExprItem.js";
import { PointsExpression } from "../../event/PointsExpression.js";
import { ReceivedPoints } from "../../event/ScoreEvent.js";
import { City } from "../../feature/City.js";
import type { Completable } from "../../feature/Completable.js";
import type { Feature } from "../../feature/Feature.js";
import { Monastery } from "../../feature/Monastery.js";
import { Road } from "../../feature/Road.js";
import { Follower } from "../../figure/Follower.js";
import type { Meeple } from "../../figure/Meeple.js";
import { AddPoints } from "../../reducers/AddPoints.js";
import { Capability } from "../Capability.js";
import type { GameState } from "../state/GameState.js";
import type { PlacedTile } from "../state/PlacedTile.js";

/** Watchtowers mini-expansion: a watchtower tile grants its meeple owner(s) a bonus when the
 *  feature the meeple sits on is scored. The bonus = `points` × (count of the subject thing in
 *  the watchtower's 3×3 neighbourhood). Subjects: coat-of-arms (pennants), monastery, city,
 *  road, meeple (adjacent followers). */
export class WatchtowerModifier extends TileModifier {
  readonly points: number;
  readonly subject: string;

  constructor(type: string) {
    super("Watchtower:" + type);
    const idx = type.indexOf("/");
    this.points = parseInt(type.substring(0, idx), 10);
    this.subject = type.substring(idx + 1);
  }
}

export class WatchtowerCapability extends Capability<void> {
  override beforeCompletableScore(state: GameState, features: Set<Completable>): GameState {
    const watchtowers = new Map<string, WatchtowerModifier>();
    const watchtowerPos = new Map<string, Position>();
    const watchtowerMeeples = new Map<string, Array<Tuple2<Meeple, FeaturePointer>>>();

    for (const t of state.getPlacedTiles()) {
      const pos = t._1;
      for (const mod of t._2.getTile().getTileModifiers()) {
        if (mod instanceof WatchtowerModifier) {
          watchtowers.set(pos.toString(), mod);
          watchtowerPos.set(pos.toString(), pos);
        }
      }
    }

    if (watchtowers.size === 0) {
      return state;
    }

    for (const t of state.getDeployedMeeples()) {
      const fp = t._2;
      const key = fp.getPosition().toString();
      if (!watchtowers.has(key)) continue;
      let onScored = false;
      for (const f of features) {
        if (f.getPlaces().contains(fp)) {
          onScored = true;
          break;
        }
      }
      if (!onScored) continue;
      let arr = watchtowerMeeples.get(key);
      if (arr === undefined) {
        arr = [];
        watchtowerMeeples.set(key, arr);
      }
      arr.push(t);
    }

    for (const [key, meeples] of watchtowerMeeples) {
      const pos = watchtowerPos.get(key)!;
      const watchtower = watchtowers.get(key)!;
      let count = 0;
      let exprName: string | null = null;

      const pts: PlacedTile[] = this.getNeigbouring(state, pos);
      let positions: Set<Position> = HashSet.empty<Position>();
      for (const pt of pts) positions = positions.add(pt.getPosition());

      switch (watchtower.subject) {
        case "coat-of-arms": {
          for (const pt of pts) {
            for (const f of pt.getTile().getInitialFeatures().values()) {
              if (f instanceof City) {
                const fp = (f as City)
                  .getPlaces()
                  .head()
                  .setPosition(pt.getPosition())
                  .rotateCW(pt.getRotation());
                const city = state.getFeature(fp) as City | null;
                if (city !== null) {
                  // City still exists (not converted to Castle)
                  if (!city.hasModifier(state, City.ELIMINATED_PENNANTS)) {
                    count += (f as City).getModifier(state, City.PENNANTS, 0);
                  }
                }
              }
            }
          }
          exprName = "pennants";
          break;
        }
        case "monastery": {
          for (const pt of pts) {
            for (const f of pt.getTile().getInitialFeatures().values()) {
              if (f instanceof Monastery) count++;
            }
          }
          exprName = "monasteries";
          break;
        }
        case "city": {
          for (const pt of pts) {
            const has = state
              .getFeatureMap()
              .get(pt.getPosition())
              .get()
              .keySet()
              .find((fp) => fp.getFeature() === (City as never))
              .isDefined();
            if (has) count++;
          }
          exprName = "cities";
          break;
        }
        case "road": {
          for (const pt of pts) {
            const has = state
              .getFeatureMap()
              .get(pt.getPosition())
              .get()
              .keySet()
              .find((fp) => fp.getFeature() === (Road as never))
              .isDefined();
            if (has) count++;
          }
          exprName = "roads";
          break;
        }
        case "meeple": {
          for (const t of state.getDeployedMeeples()) {
            if (!(t._1 instanceof Follower)) continue;
            const mpos = t._2.getPosition();
            if (Math.abs(pos.x - mpos.x) <= 1 && Math.abs(pos.y - mpos.y) <= 1) count++;
          }
          exprName = "meeples";
          break;
        }
      }

      if (count > 0) {
        const expr = new PointsExpression(
          "watchtower",
          new ExprItem(count, exprName!, count * watchtower.points),
        );
        const receivedPoints: ReceivedPoints[] = meeples.map(
          (t) =>
            new ReceivedPoints(
              expr,
              t._1.getPlayer(),
              new ScoreMeeplePositionsPointer(t._1.getDeployment(state)!, t._1.getId(), positions),
            ),
        );
        state = new AddPoints(List.ofAll(receivedPoints), true).apply(state);
      }
    }

    return state;
  }

  private getNeigbouring(state: GameState, pos: Position): PlacedTile[] {
    const out: PlacedTile[] = [];
    for (const pt of state.getAdjacentAndDiagonalTiles(pos)) out.push(pt);
    const self = state.getPlacedTile(pos);
    if (self !== null) out.push(self);
    return out;
  }
}

Capability.register(WatchtowerCapability);
